File: 0015-3sum/0015-3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
        nums.sort((a,b)=>a-b)
        let res = [];
        for(let i=0;i<nums.length && nums[i]<=0;i++){
            if(i===0 || nums[i]!==nums[i-1]){
                twoSum2(nums, i, res)
            }
        }

        return res


};

var twoSum2 = function(nums,i,res){
let left = i+1;
let right = nums.length-1

while(left<right){
    let sum = nums[i]+nums[left]+nums[right]
    if(sum<0){
        left++
    }else if(sum>0){
        right--
    }else{
        res.push([nums[i],nums[left++],nums[right--]])
        while(left<right && nums[left]===nums[left-1]){
            left++
        }
    }
}
}

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        if(len(nums)>=3) and (nums[0]==nums[len(nums)-1]) and (nums[0]==0):
            return [[0,0,0]]
        result = []
        for index in range(len(nums)-1):
            left  = index +1
            right = len(nums)-1

            while left<right:
                currsum = nums[index]+nums[left]+nums[right]
                if currsum ==0:
                    result.append([nums[index],nums[left],nums[right]])
                    left+=1
                    right-=1
                elif currsum<0:
                    left+=1
                else:
                    right-=1
    
        return [list(t) for t in set(tuple(element)for element in result)]